File: prototype/rails/src/app/domain/auth/magic_link_status.rb
module Domain
  module Auth
    # Whether a magic link may still sign someone in.
    module MagicLinkStatus
      USABLE = :usable
      EXPIRED = :expired
      CONSUMED = :consumed

      module_function

      def of(expires_at:, consumed_at:, now:)
        return CONSUMED if consumed_at
        return EXPIRED if now >= expires_at

        USABLE
      end
    end
  end
end

File: app/src/app/Domain/Auth/MagicLinkStatus.php
<?php

declare(strict_types=1);

namespace App\Domain\Auth;

use DateTimeInterface;

enum MagicLinkStatus
{
    case Usable;
    case Expired;
    case Consumed;

    public static function of(DateTimeInterface $expiresAt, ?DateTimeInterface $consumedAt, DateTimeInterface $now): self
    {
        if ($consumedAt !== null) {
            return self::Consumed;
        }

        if ($now >= $expiresAt) {
            return self::Expired;
        }

        return self::Usable;
    }
}
